app-drawer/app-drawer.html

```diff
@@ -260,70 +260,139 @@
-                  this._translateDrawer(translateDelta);
-                  this.$.scrim.style.opacity = 1 + translateDelta / this.getWidth();
-                  break;
-                case 'right':
-                  translateDelta = Math.max(0, translateDelta);
-                  this._translateDrawer(translateDelta);
-                  this.$.scrim.style.opacity = 1 - translateDelta / this.getWidth();
-                  break;
-              }
-
-              // Keep track of the last user track event with non-zero ddx (which we
-              // need to determine whether to keep the drawer open/closed) since the
-              // very last track event may have zero ddx.
-              if (event.detail.ddx !== 0) {
-                this._lastTrackDirection = event.detail.ddx;
-              }
+              this._track(event);
               break;
-
             case 'end':
-              this.$.contentContainer.style.transitionDuration = '';
-              this.$.scrim.style.transitionDuration = '';
-              this.$.scrim.style.visibility = '';
-              this.$.scrim.style.opacity = '';
-              this.transform('', this.$.contentContainer);
-
-              // Only toggle the drawer if the user intentionally swipes in one direction.
-              if (Math.abs(event.detail.dx) > 30) {
-                this.opened = (this.position === 'left' && this._lastTrackDirection > 0) ||
-                  (this.position === 'right' && this._lastTrackDirection < 0);
-              }
+              this._trackEnd(event);
               break;
           }
         }
       },
 
+      _trackStart: function(event) {
+        // Disable transitions since style attributes will reflect user track events.
+        this.$.contentContainer.style.transitionDuration = '0s';
+        this.$.scrim.style.transitionDuration = '0s';
+        this.$.scrim.style.visibility = 'visible';
+
+        if (this.opened) {
+          this._translateOffset = 0;
+        } else if (this.position == 'left') {
+          this._translateOffset = -this.getWidth();
+        } else if (this.position == 'right') {
+          this._translateOffset = this.getWidth();
+        }
+
+        this._trackEvents = [];
+      },
+
+      _track: function(event) {
+        this._translateDrawer(event.detail.dx + this._translateOffset);
+        this._trackEvents.push(event);
+      },
+
+      _trackEnd: function(event) {
+        // Save translate progress made during track events for this._rafAnimate.
+        this._translateOffset = event.detail.dx + this._translateOffset;
+        this._startTimeStamp = performance.now();
+
+        // Track handler takes precedence over scrim tap handler.
+        this.cancelDebouncer('_scrimTapHandler');
+
+        // Calculate the velocity with the first track event that is recent enough, if
+        // it exists
+        this._velocity = 0;
+        var timeLowerBound = Date.now() - 100;
```

Comment:
That won't work - if 100ms elapses between the last touch move event and touch up event, we should _not_ use it in velocity calculation, but this code will. The point is there could be 0 to infinitely many track events in the 100ms leading up to track end, and we want the first one (lowest timestamp). I could filter the list on every track event to eliminate events that are over 100ms old, but that would be expensive to do (since track events come so often).